pkg/lambci-lambda/service.go

```diff
@@ -0,0 +1,278 @@
+
+	// for lambci-lambda there is no need to create zip files
+	return rawDir, nil
+}
+
+// Install a function to the desired FaaS service. It is assumed that
+// Package() has already been called on this rawDir. The name of rawDir is
+// also the name of the function.
+func (service *lambciLambda) Install(rawDir string, env map[string]string, runtime string) error {
+
+	if runtime == "" {
+		runtime = service.defaultRuntime
+	}
+
+	// we allow no runtime for lambci as it could be provided by the container
+	if runtime != "" {
+		if _, exists := service.runtimes[runtime]; !exists {
+			return errors.Errorf("runtime '%s' does not exist in configuration", runtime)
+		}
+	}
+
+	// remove old layer
+	_, err := service.Exec(fmt.Sprintf("find %s -mindepth 1 -maxdepth 1 -exec rm -r {} +", filepath.Join(service.homeDir, runtimeDir)))
+	if err != nil {
+		return errors.Wrap(err, "error removing old layer")
+	}
+
+	// install new layer
+	if runtime != "" {
+		for _, layer := range service.runtimes[runtime] {
+			_, err := service.Exec(fmt.Sprintf("cp -r %s %s", filepath.Join(service.homeDir, layersDir, layer, "*"), filepath.Join(service.homeDir, runtimeDir)))
+			if err != nil {
+				return errors.Wrapf(err, "error installing layer '%s'", layer)
+			}
+		}
+	}
+
+	// remove old task
+	_, err = service.Exec(fmt.Sprintf("find %s -mindepth 1 -maxdepth 1 -exec rm -r {} +", filepath.Join(service.homeDir, taskDir)))
+	if err != nil {
+		return errors.Wrap(err, "error removing old task")
+	}
+
+	// install new task
+	_, err = service.Copy(filepath.Join(rawDir, "*"), filepath.Join(service.homeDir, taskDir))
+	if err != nil {
+		return errors.Wrap(err, "error installing function")
+	}
+
+	// retrieve process id of running lambda docker image
+	pid, err := service.Exec("ps ax | grep \"LAMBDA\" | grep -v entr | grep -v grep | awk \"{print $1}\"")
+	if err != nil {
+		return errors.Wrap(err, "error retrieving lambda process id")
+	}
+
+	// install new env map - this triggers the lambda function reload
+	_, err = service.Exec(fmt.Sprintf("echo -n \"%s\" > %s", Map2Lines(env), filepath.Join(service.homeDir, envFile)))
+	if err != nil {
+		return errors.Wrap(err, "error updating environment")
+	}
+
+	// wait until function reload happened in case of running lambda docker image
+	if pid != "" {
+
+		checks := 0
+		for {
+			time.Sleep(checkDelay)
+
+			newPid, err := service.Exec("ps ax | grep \"LAMBDA\" | grep -v entr | grep -v grep | awk \"{print $1}\"")
+			if err != nil {
+				return errors.Wrap(err, "error retrieving lambda process id")
+			}
+			if newPid != "" && newPid != pid {
+				break
+			}
+
+			if checks >= maxChecks {
+				return errors.Errorf("lambda function container did not reload after %v, giving up", time.Duration(checks)*checkDelay)
+			}
+			checks++
+		}
+	}
+
+	return nil
+}
+
+// Removes a function from the service. Does not affect packages.
+func (service *lambciLambda) Remove(fName string) error {
+
```

Comment:
extra newline. Also, I suppose it doesn't really matter, but why not clean up the function on the lambci side?